21.md

```diff
@@ -0,0 +1,185 @@
+- The inclusion of a `currency` implies it can be used for denomination of an amount
+- The inclusion of a `convertible currency` implies the `SERVICE` can quote and guarantee a price for a given currency
+- The `multiplier` is not guaranteed by the `SERVICE` and is subject to change
+- The `code` of a `currency` will be used as an identifier for the next request and must be unique
+- The `code` must be according to [ISO-4217](https://en.wikipedia.org/wiki/ISO_4217) if possible
+- The order of the `currencies` may be interpreted by the `WALLET` as the receiving user preference for a currency
+
+### Wallet-side second request
+
+Upon receiving the `currencies` field on the response, the `WALLET` shows the user it has the option of denominating the amount in one of the `currencies` or for the payment to be creditted as a different `currency` for the receiver.
+
+The inputs that must be gathered from the user are:
+- An optional denominating currency and amount (`CURRENCY_D` and `AMOUNT_D`)
+- An optional `convert` currency (`CURRENCY_C`)
+
+The most general case has all the parameters set.
+It will generate an invoice with the amount equivalent to `AMOUNT_D` `CURRENCY_D`, which will be converted into `CURRENCY_C` by the `SERVICE` upon payment.
+
+`<callback><?|&>amount=<AMOUNT_D>.<CURRENCY_D>&convert=<CURRENCY_C>`
+
+Each combination of parameters is valid and generates a different use case.
+- Omitting the `amount` denomination implies the invoice is for millisatoshis (base spec)
+- Omitting the `convert` implies the receiver will get BTC from the payment, no matter the `amount` denomination
+
+Note that the amount provided in all requests is always an integer number interpreted as the smallest unit of the selected `currency`. The smallest unit needs to be according to the `decimals` parameter, so the `WALLET` has all the needed information to receive input and show output properly.
+
+### Service-side second response
+
+Upon receiving a currency-denominated request from `WALLET`, the `SERVICE` must return an invoice with an amount matching the converted rate for the amount in that currency. The rate used does not need to match the `multiplier` first informed.
+
+If the `WALLET` requested an actual conversion, the `SERVICE` must provide an additional field alongside the invoice informing the guaranteed `converted` amount that will be creditted to the receiver upon payment. The `converted` amount, and therefore the conversion rate, must be guaranteed by the `SERVICE` for as long as the invoice is not expired. The `converted` amount must be denominated in the smallest unit of the currency, just like the `amount` parameter.
+
+```typescript
+type BaseResponse = {
+  pr: string,
+  routes: [],
+}
+
+type ExtendedResponse = BaseResponse & {
+  converted?: number, // Integer; Present if and only if `convert` was received.
+}
+```
+
+```diff
+{
+  "pr": "lnbc1230n1pjknkl...ju36m3lyytlwv42fee8gpt6vd2v",
+  "routes": [],
++ "converted": 123
```

Comment:
Funny enough, in UMA, I chose not to include the target amount, but did include the multiplier and some other stuff:

```typescript
"paymentInfo": {
    // The currency code of the receiving currency (eg. "USD"). This should match the requested currency in the payreq
    // request.
    "currencyCode": string,
    // Millisats per "unit" of the receiving currency (eg. 1 cent in USD). A double-precision floating point number.
    // In this context, this is just for convenience. The conversion rate is also baked into the invoice amount itself.
    // `invoice amount = amount * multiplier + exchangeFeesMillisatoshi`
    "multiplier": number,
    // Number of digits after the decimal point for the receiving currency. For example, in USD, by convention, there are
    // 2 digits for cents - $5.95. In this case, `decimals` would be 2. This should align with the currency's `decimals`
    // field in the LNURLP response. It is included here for convenience. See [UMAD-04](/uma-04-local-currency.md) for
    // details, edge cases, and examples.
    "decimals": number,
    // The fees charged (in millisats) by the receiving VASP to convert to the target currency.
    // This is separate from the multiplier rate.
    "exchangeFeesMillisatoshi": number
},
```

The fees are useful for a full visual cost breakdown on the sender side. The multiplier and decimals help infer the converted amount from the invoice amount. I do like having the wrapping struct to keep related fields grouped and allow for extension in the future without polluting the top-level structure. Just some food for thought :-). 

UMA spec link: https://github.com/uma-universal-money-address/protocol/blob/main/umad-06-payreq-response.md

Reply:
This is interesting. The reason I chose to keep just the int is to keep the protocol smaller.
The only property in this structure that is not already included is the fee, right?

The reason I didn't included it is that there's multiple ways of taking the fee, right? It could be billed in sats, or in currency (althought it's kind of the same). But the fee could be in the price (spread) or separate. Etc.

What do you think about these possibilities?

Reply:
Yeah that's fair. I just delimit the fee in mSats always because my assumption is that it's based into the cost of the invoice: `invoice amount = amount * multiplier + exchangeFeesMillisatoshi`. I like having a top-level structure to wrap payment details since I could imagine more fields being added there in the future for scenarios you're describing. With the wrapper, we wouldn't need to pollute the top level more.